src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs

```diff
@@ -71,4 +72,96 @@ private static IResourceBuilder<AzureSqlServerResource> ConfigureDefaults(this I
+                // the principalType.
+                sqlServer.AssignParameter(x => x.Administrators.PrincipalType, construct.PrincipalTypeParameter);
+
+                var sqlFirewall = new SqlFirewallRule(construct);
+                sqlFirewall.AssignProperty(x => x.StartIPAddress, "'0.0.0.0'");
+                sqlFirewall.AssignProperty(x => x.EndIPAddress, "'255.255.255.255'");
+            }
+
+            List<SqlDatabase> sqlDatabases = new List<SqlDatabase>();
+            foreach (var databaseNames in builder.Resource.Databases)
+            {
+                var databaseName = databaseNames.Value;
+                var sqlDatabase = new SqlDatabase(construct, sqlServer, databaseName);
+                sqlDatabases.Add(sqlDatabase);
+            }
+
+            sqlServer.AddOutput(x => x.FullyQualifiedDomainName, "sqlServerFqdn");
+
+            if (configureResource != null)
+            {
+                configureResource(construct, sqlServer, sqlDatabases);
+            }
+        };
+
+        var resource = new AzureSqlServerConstructResource(builder.Resource, configureConstruct);
+        var azureSqlDatabase = builder.ApplicationBuilder.CreateResourceBuilder(resource);
+        azureSqlDatabase.WithParameter(AzureBicepResource.KnownParameters.PrincipalId)
+                        .WithParameter(AzureBicepResource.KnownParameters.PrincipalName)
+                        .WithManifestPublishingCallback(resource.WriteToManifest);
+
+        if (builder.ApplicationBuilder.ExecutionContext.IsRunMode)
+        {
+            azureSqlDatabase.WithParameter(AzureBicepResource.KnownParameters.PrincipalType);
+        }
+
+        if (useProvisioner)
+        {
+            // Used to hold a reference to the azure surrogate for use with the provisioner.
+            builder.WithAnnotation(new AzureBicepResourceAnnotation(resource));
+            builder.WithConnectionStringRedirection(resource);
+
+            // Remove the container annotation so that DCP doesn't do anything with it.
+            if (builder.Resource.Annotations.OfType<ContainerImageAnnotation>().SingleOrDefault() is { } containerAnnotation)
+            {
+                builder.Resource.Annotations.Remove(containerAnnotation);
+            }
+        }
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Configures SQL Server resource to be deployed as Azure SQL Database (server).
+    /// </summary>
+    /// <param name="builder">The builder for the SQL Server resource.</param>
+    /// <param name="configureResource">Callback to customize the Azure resources that will be provisioned in Azure.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<SqlServerServerResource> PublishAsAzureSqlDatabaseConstruct(this IResourceBuilder<SqlServerServerResource> builder, Action<ResourceModuleConstruct, SqlServer, IEnumerable<SqlDatabase>>? configureResource = null)
```

Comment:
Feels like we should expose the database list on a context object or something, or maybe not at all? Can you inspect these from the sql server or construct? Is the idea to pass any auto created resource to the callback?

Reply:
So I think we'll probably end up revisiting this. I like the idea of having a context, but I don't like the idea of having a context type for every single `AddAzureX`/`PublishAsAzureX` method.

So my cheap solution for now is to just add arguments to the function for things that people might want to customize. The construct gives them access to everything, but the extra arguments are mostly for ergonomics.

Reply:
If yes, then each should have a context (what if we add more default resources in the future??).